src/engine/guidance/collapse_scenario_detection.cpp

```diff
@@ -81,8 +95,11 @@ bool basicCollapsePreconditions(const RouteStepIterator first,
                                      hasRoundaboutType(second->maneuver.instruction) ||
                                      hasRoundaboutType(third->maneuver.instruction);
 
+    const auto contains_bad_uturn =
+        !isValidUturnCombination(first, second) && !isValidUturnCombination(second, third);
+
     // require modes to match up
-    return !has_roundabout_type && haveSameMode(*first, *second, *third);
+    return !has_roundabout_type && haveSameMode(*first, *second, *third) && !contains_bad_uturn;
```

Comment:
General note here: should `basicCollapsePreconditions(first, second, third)` run `basicCollapsePreconditions(first, second)` on its `first, second` and `second, third` argument here? If not, why?